docs/build/voting-dapp-tutorial.md

```diff
@@ -0,0 +1,428 @@
+---
+title: dApp Tutorial
+description: Build a complete dApp on Kadena
+---
+
+import PageRef from '@components/PageRef'
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Complete dApp Tutorial
+
+Build a complete on-chain voting dApp
+
+---
+
+## Overview
+
+In this tutorial you will build a complete dApp, smart contracts and frontend that allows users to submit a vote. You will also deploy your own gas station that will be used to pay for the gas needed to interact with your dApp.
+
+## Environment Setup
+
+### Install Pact
+
+**Using Brew**
+
+```clojure
+brew update
+brew install kadena-io/pact/pact
+```
+**Build from source**
+- follow the instructions [here](https://github.com/kadena-io/pact#building-from-source)
+
+**Verifying Installation**
+
+Type `pact` in a terminal and try out some commands:
+
+```clojure
+$ pact
+pact> (+ 1 2)
+3
+pact> (+ "hello, " "world")
+"hello, world"
+pact> (< 0 1)
+true
+```
+
+### Create Project Structure
+
+Copy paste the instructions below to create a basic project structure:
+
+```clojure
+mkdir my-pact-project && cd my-pact-project
+mkdir pact
+mkdir app
+```
+
+### Pact Http Server
+
+Create a `config.yaml` file for pact http server. We will later use it simulate blockchains transactions locally.
+
+For more details about pact http server check <a href="/blog/deploy-to-a-local-server">this</a> tutorial.
+
+```clojure
+touch config.yaml
+```
+Copy the code below and save the file:
+
+```clojure
+# Config file for pact http server. Launch with `pact -s config.yaml`
+
+# HTTP server port
+port: 8080
+
+# directory for HTTP logs
+logDir: log
+
+# persistence directory
+persistDir: log
+
+# SQLite pragmas for pact back-end
+pragmas: []
+
+# verbose: provide log output
+verbose: True
+```
+
+### Atom IDE (Optional)
+
+You can use any text editor to write Pact but if you prefer the benefits of an IDE, "language-pact" is a package for [Atom](https://atom.io) that provides syntax highlighting and linting.
+
+To install it, go to Preferences -> Packages -> Type "language-pact" and click the Install button.
+
+![atom-language-pact](/img/docs/voting-dapp/atom-language-pact.png)
+
+## Smart Contracts
+
+Our voting app should allow you to submit a vote while not allowing an address to vote more than once.
+Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying gas.
+
+### Voting
+
+Let's start to implement our voting contract.
+
+Assuming you are still in your project directory, create the `vote.pact` and `vote.repl` files:
+
+```clojure
+touch pact/vote.pact pact/vote.repl
+```
+
+`vote.pact` is the contract source code while in `vote.repl` we'll write tests.
+
+Copy the following code in the `vote.pact` file:
+
+```clojure
+(define-keyset 'vote-admin-keyset)
+
+(module simple-vote 'vote-admin-keyset
```

Comment:
Have one of those "Warning boxes" here that points out that module names need to be unique, and as such, they will need to choose a different module name when they go to actually deploy to testnet. Also maybe use the free namespace from the start?

Reply:
I added it and I also mention this in the deployment section. Regarding the namespace, I mentioned it at the deployment step cause otherwise I'd have to say that you need to comment it while you're running tests in REPL or I need to write more code for that namespace definition for testing purposes and I thought it's too much overload at start.

Reply:
Rules like this should be in some sort of pact rules part where we again ca link towards for now I think this is pretty nice